modules/dnn/test/test_model.cpp

```diff
@@ -806,6 +807,85 @@ TEST_P(Test_Model, TextDetectionByEAST)
     );
 }
 
+TEST_P(Test_Model, YOLOv7)
+{
+    std::string weightPath = _tf("onnx/models/yolov7_not_simplified.onnx");
+    std::string imgPath = _tf("dog_orig_size.jpg");
+
+    Size targetSize{640, 640};
+    float conf_threshold = 0.3;
+    float iou_threshold = 0.5;
+
+    // Reference, which is collected with input size of 640x640
+    std::vector<int> refClassIds{1, 16, 7};
+    std::vector<float> refScores{0.9624768f, 0.95940655f, 0.8693003f};
+    // [x1, y1, x2, y2] x 3
+    std::vector<Rect2d> refBoxes{Rect2d(105.99881f, 150.2002f,  472.52933f, 466.508f),
+                                  Rect2d(109.99532f, 246.13585f, 259.82266f, 600.7827f),
+                                  Rect2d(386.13947f, 83.04132f,  575.3909f,  189.83206f)};
+
+    Mat img = imread(imgPath);
+    Mat inp = blobFromImage(img, 1/255.0, targetSize, Scalar(0, 0, 0), true, false);
+
+    Net net = readNet(weightPath);
+
+    net.setInput(inp);
+    std::vector<Mat> outs;
+    net.forward(outs, net.getUnconnectedOutLayersNames());
+
+    Mat preds = outs[3].reshape(1, outs[3].size[1]); // [1, 25200, 85]
+
+    // Retrieve
+    std::vector<int> classIds;
+    std::vector<float> confidences;
+    std::vector<Rect2d> boxes;
+    // each row is [cx, cy, w, h, conf_obj, conf_class1, ..., conf_class80]
+    for (int i = 0; i < preds.rows; ++i)
+    {
+        // filter out non objects
+        float obj_conf = preds.row(i).at<float>(4);
+        if (obj_conf < conf_threshold)
+            continue;
+
+        // get class id and conf
+        Mat scores = preds.row(i).colRange(5, preds.cols);
+        double conf;
+        Point maxLoc;
+        minMaxLoc(scores, 0, &conf, 0, &maxLoc);
+        conf *= obj_conf;
+        if (conf < conf_threshold)
+            continue;
```

Comment:
To my understanding, recent object detection models (e.g. [YOLOX in opencv_zoo](https://github.com/opencv/opencv_zoo/pull/86/files#diff-e00ac7b5ef6f02331aac9d5b854ce37b7f6c8cbaa60aeb376a8c5bec75b26f07R70)) are trained with predicting IoU between anchors and ground truth boxes ([related lines in YOLOv7 repo](https://github.com/WongKinYiu/yolov7/blob/44d8ab41780e24eba563b6794371f29db0902271/utils/loss.py#L468-L472)). In this case, the final confidence should be calculated by multiplying the class confidence and the predicted IoU (which is named `obj_conf` in my code).

In short, models without such a training scheme should not do this multiplication. So I do not think it makes sense if we do this for region/detection_output layer, considering compatibility for old object detection models.

By the way, we should consider support next gen NMS algorithms (CIoU, DIoU ...) since they are quite popular among recent object detection models.